packages/x3d-serializer/index.js

```diff
@@ -1,120 +1,154 @@
-// import xmldom from 'xmldom'
-const xmldom = require('xmldom')
-const { ensureManifoldness } = require('@jscad/io-utils')
+/*
+JSCAD Object to X3D (XML) Format Serialization
+
+## License
+
+Copyright (c) 2018 JSCAD Organization https://github.com/jscad
+
+All code released under MIT license
+
+Notes:
+1) CAG conversion to:
+     none
+2) CSG conversion to:
+     IndexedTriangleSet with Coordinates and Colors
+3) Path2D conversion to:
+     none
+
+TBD
+1) gzipped is also possible; same mime type, with file extension .x3dz
+*/
+
+//const {ensureManifoldness} = require('@jscad/io-utils')
+const {isCSG} = require('@jscad/csg')
+const {ensureManifoldness} = require('@jscad/io-utils')
+
+const stringify = require('onml/lib/stringify')
+
+// http://www.web3d.org/x3d/content/X3dTooltips.html
+// http://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html#Meshes
+// https://x3dgraphics.com/examples/X3dForWebAuthors/Chapter13GeometryTrianglesQuadrilaterals/
 
 const mimeType = 'model/x3d+xml'
 
-const XMLSerializer = xmldom.XMLSerializer
-// NOTE: might be useful :https://github.com/jindw/xmldom/pull/152/commits/be5176ece6fa1591daef96a5f361aaacaa445175
-
-function serialize (CSG, options) {
-  options && options.statusCallback && options.statusCallback({progress: 0})
-  CSG = ensureManifoldness(CSG)
-  const DOMImplementation = typeof document !== 'undefined' ? document.implementation : new xmldom.DOMImplementation()
-  // materialPolygonLists
-  // key: a color string (e.g. "0 1 1" for yellow)
-  // value: an array of strings specifying polygons of this color
-  //        (as space-separated indices into vertexCoords)
-  var materialPolygonLists = {}
-  // list of coordinates (as "x y z" strings)
-  var vertexCoords = []
-  // map to look up the index in vertexCoords of a given vertex
-  var vertexTagToCoordIndexMap = {}
+/** Serialize the give objects to X3D (xml) format.
+ * @param {Object} [options] - options for serialization
+ * @param {Object|Array} objects - objects to serialize as X3D
+ * @returns {Array} serialized contents, X3D format
+ */
+const serialize = (options, ...objects) => {
+  const defaults = {
+    statusCallback: null,
+    unit: 'millimeter', // millimeter, inch, feet, meter or micrometer
+    color: [0, 0, 1, 1.0], // default colorRGBA specification
```

Comment:
does this override the colors already defined on the shapes themselves ?
in which case I am not sure this should be here: I did something similar for the new 3D viewer, and sometimes figuring out which color is applied from where can get confusing
Also I do not thing a serializer should modify that kind of input (csg) data ?

Reply:
It supplies a default color only. It doesn’t override colors in objects. This is required as far as I know.